### PR TITLE
Generics support

### PR DESCRIPTION
**Pull request by majakusber**
Revert of PR #1411, adding fixes for parsing edge cases of generics, adding unit tests

### CODE REVIEW

File: regression/cbmc-java/enum1/test.desc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (thk123) [inline regression/cbmc-java/enum1/test.desc]: These hopefully don't need to be turned off any more as it shouldn't crash when they are found.

**Comment by thk123** [comment]
I suggest reviewers look at [this diff](https://github.com/diffblue/cbmc/pull/1434/files/e36d2cd25d121654c22df8c097ad10b017245e45..9b4662d3bc41db55bcca7948b726bcb1b295fd3c) as the original PR (#1322) is reapplied whole sale and has already been reviewed

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 2]
```diff
@@ -94,7 +94,7 @@ void java_bytecode_convert_classt::convert(const classt &c)
   }
 
   java_class_typet class_type;
-  if(c.signature.has_value())
+  if(c.signature.has_value() && c.signature.value()[0]=='<')
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: @mgudemann This was to fix parsing classes that extend from generic classes

File: src/java_bytecode/java_types.cpp [hunk 3]
```diff
@@ -261,7 +285,8 @@ typet java_type_from_string(
       const typet subtype=
         java_type_from_string(src.substr(1, std::string::npos), class_name);
       if(subtype_letter=='L' || // [L denotes a reference array of some sort.
-         subtype_letter=='[')   // Array-of-arrays
+         subtype_letter=='[' || // Array-of-arrays
+         subtype_letter=='T')   // Array of generic types
```
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: @mgudemann Deals with reference types that are of the form `T []`

File: src/java_bytecode/java_types.cpp [hunk 4]
```diff
@@ -389,6 +414,55 @@ typet java_type_from_string(
 
       return java_reference_type(symbol_type);
     }
+  case '*':
+  {
```
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: Probably these three cases (`'*'`, `'-'`, `'+'`) should be handled by throwing an exception rather than this ugly partial handling.

File: src/java_bytecode/java_types.cpp [hunk 5]
```diff
@@ -438,13 +512,16 @@ std::vector<typet> java_generic_type_from_string(
...
-  /// the class signature is of the form <TX:Bound_X;TY:Bound_Y;>base_class
-  size_t signature_end=src.find('>');
+  /// the class signature is of the form <TX:Bound_X;:BoundZ;TY:Bound_Y;>
+  size_t signature_end=find_closing_delimiter(src, 0, '<', '>');
   INVARIANT(
     src[0]=='<' && signature_end!=std::string::npos,
     "Class signature has unexpected format");
   std::string signature(src.substr(1, signature_end-1));
 
+  if(signature.find(";:")!=std::string::npos)
```
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: @mgudemann this is to resolve when there are multiple bounds on a type

File: unit/java_bytecode/java_bytecode_parse_generics/Bar.java [hunk 6]
```diff
@@ -0,0 +1,2 @@
+class Bar extends Foo
+{}
```
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/Bar.java]: Nit: this file (and others) are missing a new line at the end of the file (and others). You can make sure CLion inserts these using Settings > Editor > General > Ensure Line Feed at end of file on save

**Review by thk123 — COMMENTED** [review]
Some small nits - I won't approve as I had a hand in writing some of this.

File: src/java_bytecode/java_types.cpp (hunk 4 above)
Comment (chrisr-diffblue) [reply src/java_bytecode/java_types.cpp]: I'd agree with @thk123

File: src/java_bytecode/java_types.cpp [hunk 7]
```diff
@@ -389,6 +414,55 @@ typet java_type_from_string(
...
+  {
+#ifdef DEBUG
+    std::cout << class_name << std::endl;
+#endif
+    // parse name of type variable
+    INVARIANT(src[src.size()-1]==';', "Generic type name must end on ';'.");
+    PRECONDITION(!class_name.empty());
+//    INVARIANT(
+//      src.substr(1, src.size()-2)=="*",
+//      "Wildcard types do not have a name");
```
Comment (chrisr-diffblue) [inline src/java_bytecode/java_types.cpp]: I assume these commented out INVARIANT's are intended to be re-instated once the generics support improves - if so, is it worth marking them all with FIXME or TODO comments as well to make that clearer? @thk123 is there any common practice here to include a link to a Jira or PR for in the comment to indicate what the work is covered under?

**Review by chrisr-diffblue — COMMENTED** [review]
Looks pretty good, but like @thk123 a couple if minor nits. Happy to approve if they are addressed, or there's a good explanation why the PR is fine as it is.

File: unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    x.t = assign;
+  }
+
+  // It is not legal Java to specify both an upper and lower bound
+  // public static void processBoundSuperClassGeneric(SimpleGeneric<? extends Object super Foo> x, Foo assign) {
+  //   x.t = assign;
+  // }
+
+  // Test a wild card generic bound by a class
+  // public static void processBoundClassGenericDoubleBound(SimpleGeneric<? extends Foo & BasicInterface> x) {
```
Comment (mgudemann) [inline unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:31]: if wildcards can have multiplle bounds, we should have this as test

File: unit/java_bytecode/java_bytecode_parse_generics/GenericFunctions.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  public static <T extends BasicInterface> void processUpperBoundInterfaceGeneric(SimpleGeneric<T> x) {
+    assert(x.t.getX() == 4);
+  }
+
+  // Test a wild card generic bound by a class
+  public static <T extends Foo> void processUpperBoundClassGeneric(SimpleGeneric<T> x) {
+    assert(x.t.getX() == 4);
+  }
+
+  public static <T extends Foo & BasicInterface> void processDoubleUpperBoundClassGeneric(SimpleGeneric<T> x) {
```
Comment (mgudemann) [inline unit/java_bytecode/java_bytecode_parse_generics/GenericFunctions.java]: maybe add test with two interface bounds and no class, too

File: src/java_bytecode/java_types.cpp [hunk 10]
```diff
@@ -158,13 +167,54 @@ exprt java_bytecode_promotion(const exprt &expr)
...
+  //    Ljava/lang/Object;
+  //  - type variables are similar to object types but have the prefix 'T'
   switch(src[0])
   {
+  case '<':
+    {
+      // The method signature can optionally have a collection of formal
+      // type parameters (e.g. on generic methods on non-generic classes
+      // or generic static methods). For now we skip over this part of the
+      // signature and continue parsing the rest of the signature as normal
```
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp:203]: maybe give a concrete example here what would be skipped
could this '<' case and the '<' case inside 'L' - object name parsing be merged ?

File: src/java_bytecode/java_types.cpp [hunk 11]
```diff
@@ -233,14 +302,109 @@ typet java_type_from_string(const std::string &src)
...
+    return java_generic_parametert(
+      type_var_name,
+      java_type_from_string("Ljava/lang/Object;").subtype());
+  }
   case 'L':
     {
       // ends on ;
       if(src[src.size()-1]!=';')
         return nil_typet();
       std::string class_name=src.substr(1, src.size()-2);
```
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: as discussed, this shadows the `class_name` parameter and should be moved down to where this string is used (after the generic signature case)

File: src/java_bytecode/java_types.cpp (hunk 4 above)
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: For the record - I should say it was me who wrote the  "ugly partial handling" :stuck_out_tongue:

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 12]
```diff
@@ -93,6 +94,33 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+      for(const typet &t : generic_types)
+      {
+        generic_class_type.generic_types()
+          .push_back(to_java_generic_parameter(t));
+      }
+      class_type=generic_class_type;
+    }
+    catch(unsupported_java_class_siganture_exceptiont)
+    {
+      // Do nothing: we don't support parsing for example double bounded or
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: maybe here we should fall back to using the descriptor, and emit a warning, in order to degrade gracefully

**Review by mgudemann — CHANGES_REQUESTED** [review]

File: unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java (hunk 8 above)
Comment (thk123) [reply unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:31]: Sorry comment not clear - wildcards can not have multiple bounds - this java code doesn't compile :)

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 12 above)
Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: My understanding is the `descriptor` won't have any generic type information so isn't applicable here (We are only trying to fill out generic information for the class)

Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Adding a warning is a good idea however.

File: src/java_bytecode/java_types.cpp (hunk 10 above)
Comment (thk123) [reply src/java_bytecode/java_types.cpp:203]: No - this should only occur for functions that look like:

```java
void <T> foo(T t)
```

etc (i.e. functions that are generic as opposed to part of a generic class). Perhaps a more robust way of dealing with this is to throw and exception and fall back to the descriptor in this case (since this is broadly what this is doing anyway)

**Review by thk123 — CHANGES_REQUESTED** [review]
Blocking until we've resolved issues with the TG repo

**Comment by majakusber** [comment]
Thank you everyone for the reviews, I've now pushed all the changes according to your notes. Let me know what you think.

File: src/java_bytecode/java_types.cpp (hunk 10 above)
Comment (thk123) [reply src/java_bytecode/java_types.cpp:203]: Proposed comment extension:
```C++
// So for example, the following java method:
// static void <T, U> foo(T t, U u, int x);
// Would have a signature that looks like:
// <T:Ljava/lang/Object;U:Ljava/lang/Object;>(TT;TU;I)V
// So we skip all inside the angle brackets and parse the rest of the string:
// (TT;TU;I)V
```

Comment (majakusber) [reply src/java_bytecode/java_types.cpp:203]: Thank you Thomas, adding it in..

**Review by thk123 — APPROVED** [review]

**Comment by thk123** [comment]
Issues with TG repo have been resolved: just need this PRs conflict resolved and diffblue/test-gen#1080 to pass.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+      THEN("The method should have generic return type")
+      {
+        const symbolt &method_symbol=
+          new_symbol_table
+            .lookup("java::generics$bound_element.f:()Ljava/lang/Number;");
+        const typet &symbol_type=method_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
```
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: This test case doesn't actually test what is stated in the 'THEN' clause - and I think this functionality has broken. I've just rebased my work on top of this PR, and one of my unit tests is failing because of this. Previously, you could call `is_java_generic_parameter(code.return_type())` and get the answer `true` for this function. I'm not finding this is returning `false` instead, and doing 'method_symbol.show(std::cout)` is showing that the return type for this method is not marked as being generic in any way.

File: src/java_bytecode/java_types.cpp [hunk 14]
```diff
@@ -158,13 +167,61 @@ exprt java_bytecode_promotion(const exprt &expr)
...
+      // or generic static methods). For now we skip over this part of the
+      // signature and continue parsing the rest of the signature as normal
+      // So for example, the following java method:
+      // static void <T, U> foo(T t, U u, int x);
+      // Would have a signature that looks like:
+      // <T:Ljava/lang/Object;U:Ljava/lang/Object;>(TT;TU;I)V
+      // So we skip all inside the angle brackets and parse the rest of the
+      // string:
+      // (TT;TU;I)V
+      size_t closing_generic=find_closing_delimiter(src, 0, '<', '>');
```
Comment (NlightNFotis) [inline src/java_bytecode/java_types.cpp:211]: There is a particular problem with this function when passed a `src` string with length `1` only, and it results in a huge recursive call stack, and makes the program result in a `Segmentation Violation` fault. I shall send tests and perhaps a fix in a bit.

**Review by NlightNFotis — CHANGES_REQUESTED** [review]

**Review by mgudemann — APPROVED** [review]
LGTM, please squash some of the commits before merge

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp (hunk 13 above)
Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Thanks for this, this has also been failing the CI tests. The reason is that the return type of the lookup method for symbol table has changed (commit #1444 ) - now it's a reference wrapper around symbolt instead of symbolt.

Comment (chrisr-diffblue) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Yes, that is part of the failure (it's broken my work too :-) ), but my broader point is that this test case doesn't *actually* test what it claims to test - it claims to test that it parses methods with a generic return type, but it only does that partially - it checks that it can parse the method, and that the symbol for the method exists, and is an ID_code object - but it *doesn't* check that the return type of that `code_typet` actually records that it is generic. So, for instance, we could be parsing the the method and then throwing away all the generic information and this test as it is currently written would still be passing. This appears to have happened at some point, because when my code was based off @thk123 branch, the `is_java_generic_parameter(code.return_type())` was correctly returning true for this case, whereas with this branch I'm seeing that the generic information is no longer stored for the return type.

Comment (thk123) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Yes - this is the mess with descriptors of methods being used. As discussed I think before we do [TG-865](https://diffblue.atlassian.net/browse/TG-865) it won't work. The THEN should be updated in a similar way to the TODO directly below.
